### PR TITLE
Introduce new GH workflow for releases

### LINKED ISSUE #1901: Automatically create a release once a release branch is merged
## Problem

At the moment, releases have to be triggered manually by pushing a tag.

## Goal

Less manual interaction for releases.

## Scope

* Automate the release as described below
* Document what has to be done to release a new version of cnd. The goal of this document should be that everyone on the team can do a release. 
  * Create a `RELEASING.md` document in the repository. 

## Recommendation

Building on top of https://github.com/comit-network/comit-rs/issues/1900, we can safely trigger a release once any branch that starts with `release/` is merged.

Create a GitHub workflow that:

- listens for `push` events on `master` (with https://github.com/comit-network/comit-rs/issues/1900 implemented, this only happens when `release/` branches are merged)
- build linux and macos release binaries (in parallel)
- create a release on github
- attach the binaries to the release
- create a docker image that contains the release binary
- push the docker image to dockerhub

Note: Should be feasible without any custom made GitHub action.

You can take the following workflow as a blueprint: https://github.com/thomaseizinger/github-action-gitflow-release-workflow/blob/dev/.github/workflows/publish-new-release.yml

### PR DESCRIPTION
**Pull request by thomaseizinger**
I've tested this to my best ability by creating dummy releases on push events to this branch. You can see the final successful execution here: https://github.com/comit-network/comit-rs/actions/runs/45203040

I did have to make a bunch of changes afterwards to use the actual variables etc.

Please focus your review on finding mistakes in that regards:

- referencing outputs that don't exist
- referencing env variables that don't exist
- etc

The overall workflow works like this:

- macos and ubuntu runners build artifacts in parallel. they create archives and upload them to github artifacts
- two ubuntu jobs trigger after this: releasing the docker image and creating the github release
- the docker job downloads only the ubuntu artifact, builds the image and publishes it on dockerhub
- the GH release job creates a new GH release, downloads the two artifacts and uploads them to the release. I went with the official upload-release-asset action because I had to separate the creation of the release from the uploading. I used my own fork of the create-release action for reasons stated in the readme: https://github.com/thomaseizinger/create-release#features

Fixes #1901.

### CODE REVIEW

File: RELEASING.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+Releases of comit-rs are mostly automated.
+Here is what you need to know:
+
+The repository uses the GitFlow branching model.
+Hence, releases are started by branching a release-branch (`release/x.y.z`) of `dev`.
+
+1. Start with creating said release branch locally.
+2. Update the changelog to reflect the newest release.
+3. Bump necessary versions in the manifest files.
```
Comment (D4nte) [inline RELEASING.md]:
```suggestion
3. Bump necessary versions in the manifest files.
4. Build to ensure the lock files are updated.
```

File: RELEASING.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+4. Commit and push your changes
+5. Create a pull request targeting the __master__ branch.
+6. Get approvals and merge it.
+
+From here, [the automation](./.github/workflows/publish-new-release.yml) takes over and:
+
+1. Build the release artifacts for Linux and MacOS.
+2. Pushes a new tag of the [comitnetwork/cnd](https://hub.docker.com/repository/docker/comitnetwork/cnd) docker image.
+3. Tags the merge commit and creates a release on GitHub with the binaries attached.
```
Comment (D4nte) [inline RELEASING.md]:
```suggestion
3. Tags the merge commit and creates a release on GitHub with the binaries attached.
4. Opens a PR to merge back to _dev_ branch, do ensure it gets merged.
```
Or do the job merge the PR automatically?

Comment (thomaseizinger) [reply RELEASING.md]: It just opens it, good catch!

File: .github/workflows/publish-new-release.yml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+          name: ${{ matrix.os }}-release-archive
+          path: ${{ steps.create-archive.outputs.archive }}
+
+  create_docker_image:
+    if: github.event.pull_request.merged == true && startsWith(github.event.pull_request.head.ref, 'release/') # only merged release branches must trigger this
+    name: Create and publish Docker image
+    needs: build_binary
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout merge commit
```
Comment (bonomat) [inline .github/workflows/publish-new-release.yml]: Do you need to checkout the merge commit here? Further down you only download the release, extract it and build the container.

File: RELEASING.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Releasing
+
+Releases of comit-rs are mostly automated.
+Here is what you need to know:
+
+The repository uses the GitFlow branching model.
+Hence, releases are started by branching a release-branch (`release/x.y.z`) of `dev`.
```
Comment (luckysori) [inline RELEASING.md]: :wrench: 
```suggestion
Hence, releases are started by branching a release-branch (`release/x.y.z`) off `dev`.
```

File: .github/workflows/publish-new-release.yml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create-release.outputs.upload_url }}
+          asset_path: ./macos-release-archive/${{ steps.detect-archives.outputs.macos-archive }}
+          asset_name: ${{ steps.detect-archives.outputs.macos-archive }}
+          asset_content_type: application/gzip
+
+  merge_release_into_dev:
```
Comment (bonomat) [inline .github/workflows/publish-new-release.yml]: Shall we automate this as well to ensure the timeliness?
I see two ways: 
1) comment `bors r+` right after you created the PR
2) just merge it with another action :)

What do you think?

File: RELEASING.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+1. Update the changelog to reflect the newest release.
+1. Bump necessary versions in the manifest files.
+1. Build to ensure the lock files are updated.
+1. Commit and push your changes
+1. Create a pull request targeting the __master__ branch.
+1. Get approvals and merge it.
+
+From here, [the automation](./.github/workflows/publish-new-release.yml) takes over and:
+
+1. Build the release artifacts for Linux and MacOS.
```
Comment (luckysori) [inline RELEASING.md]: :wrench: 
```suggestion
1. Builds the release artifacts for Linux and MacOS.
```

File: .github/workflows/publish-new-release.yml (hunk 5 above)
Comment (thomaseizinger) [reply .github/workflows/publish-new-release.yml]: Yes, we can add mergify rules for that at some point :)

File: .github/workflows/publish-new-release.yml (hunk 3 above)
Comment (thomaseizinger) [reply .github/workflows/publish-new-release.yml]: I need the Dockerfile, which is in the repository :)

**Pull request merged**